mantidproject/mantid
# Add caching to absorptioncorrutil for SNSPowderReduction.

**Description of work.**

Adding one extra caching step in existing absorption calculation function so that multiple call to `SNSPowderReduction` can utilize cached absorption correction workspace to speed up the overall processing time of multiple runs.

> sibling PR to `ornl-next`: #30491 

**To test:**
Use the following code on the analysis cluster (or mount the data drive in the testing env) 

```python
SNSPowderReduction("/SNS/PG3/IPTS-2767/nexus/PG3_46214.nxs.h5",
                   CalibrationFile="/SNS/PG3/shared/CALIBRATION/2020_1_11A_CAL/PG3_PAC_HR_d46168_2020_05_06.h5",
                   CharacterizationRunsFile="/SNS/PG3/shared/CALIBRATION/2020_2_11A_CAL/PG3_char_2020_01_04_PAC_limit_1.4MW.txt,/SNS/PG3/shared/CALIBRATION/2020_2_11A_CAL/PG3_char_2020_05_06-HighRes-PAC_1.4 MW.txt",
                   Binning=-0.001,
                   SaveAs="nexus",
                   TypeOfCorrection="SampleAndContainer",
                   SampleFormula="La-(B11)5.94-(B10)0.06",
                   MeasuredMassDensity=2.36,
                   ContainerShape="PAC06",
                   OutputFilePrefix='SC_absorption_',
                   OutputDirectory=".",
                   CacheDir=".",
                   )
```

- First call will run the reduction as before, with additional workspace being saved to the memory as well as the specified cache directory.
- To check the effect of
  * __Hard disk cache__: clear the workbench memory space, then rerun `SNSPowderReduction` again with the same code block above.  The overall time should be reduced to about half due to bypassing the calculation of absorption correction via caching.
  * __Memory cache__: remove all other workspaces other than the two with `_ass` and `_acc` (or `_assc` and `_ac` if using `FullPaalmanPings` method), then rerun `SNSPowderReduction` again with the same code block above.  The overall time should be reduced to about half due to bypassing the calculation of absorption correction via caching.




Fixes #30489. 





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
